### ISSUE TITLE
CPP Microservices SegFault

### ISSUE THREAD

**Issue by DmitryLyakh** [ORNL-QCI/exatn#17]
In src/runtime/tensor_runtime.cpp (lines 10-11), the TensorRuntime constructor is supposed to created an instance of TensorGraphExecutor and TensorNodeExecutor via the call to exatn::getService<> from CPP Microservices. However it results in SegFault in all CTEST tests, which makes me believe something got screwed up again with the CPP Microservices usage. The lines 10-11 are currently commented out, but we need to uncomment them and fix the SegFaults.

**Comment by DmitryLyakh** [comment]
In fact, some tests report the following before the SegFault:
ExaTN is not initialized: Please execute exatn::initialize() before using its API.

**Comment by amccaskey** [comment]
This will happen if you try to get reference to a service before the ServiceRegistry has been initialized. This is what happened in Issue #7. You can't call getService from another service implementation until after ServiceRegistry::initialize() gets called and returns. 

ServiceRegistry::initialize() is going to load all plugin libraries and call the Bundle Activator start() method, which is going to call the constructor of the specific service it is activating. If during this start() invocation, a call to getService() occurs, you'll get errors like this because initialize() hasn't finished and not all services have been loaded.

**Comment by amccaskey** [comment]
I would suggest adding a tensor runtime initialize() method that has to be invoked after construction, which gets reference to those services. Or just check before its used if the service is nullptr, and if it is call getService then

**Comment by DmitryLyakh** [comment]
Alex, I am not sure I understand these subtleties :) Are you saying that I cannot call getService<TensorGraphExecutor> and getService<TensorNodeExecutor> inside TensorRuntime::TensorRuntime constructor at all? I need to construct a TensorRuntime object with either default or provided executors (line 10-11 in tensor_runtime.cpp). What should I add for this to work?

**Comment by amccaskey** [comment]
Yea that's correct. My solution to this would be to add TensorRuntime::initialize(), and implement it to do the same thing you are trying to do right now in the constructor. Downside there is that you would then have to construct the TensorRuntime and call initialize() to get it working properly, but if you document that then it is fine. 

But if you don't want to add the extra call, you could do something like 

```cpp
if ( ! graph_executor_ ) {
  graph_executor = exatn::getService<TensorGraphExecutor>(graph_executor_name);
}```

in one of the methods where you need graph_executor_, then the first time it is going to be used, if it is not allocated, it will be created.

**Comment by DmitryLyakh** [comment]
This is a bit ugly. I am still puzzled why I cannot call getService inside a constructor. Is it a general rule that getService cannot be called inside a constructor? And when exactly and by whom serviceRegistry::initialize() is called?

**Comment by DmitryLyakh** [comment]
In other words, can I leave the constructor as it is (calling getService as needed), but make sure somehow that serviceRegistry::initialize() had already been called and completed prior to the moment of TensorRuntime construction?

**Comment by amccaskey** [comment]
You cannot getService() from the constructor of another service. 

Actually I just realized that here you are calling getService from a non-service (just TensorRuntime). YOu still have to have called exatn::initialize() before you ever call getService. But as exatn::initialize() has been called, I would imagine that it should work.

**Comment by DmitryLyakh** [comment]
Yes, that's exactly the problem. TensorRuntime is not a service, that's why I got confused why calling getService() from inside its constructor causes SegFault due to uninitialized status of serviceRegistry.
